server/stream.go

```diff
@@ -1780,6 +1780,27 @@ func (mset *stream) updateWithAdvisory(config *StreamConfig, sendAdvisory bool)
 	// Now update config and store's version of our config.
 	mset.cfg = *cfg
 
+	// Check for changes to the retention policy. If we are moving to interest-based
+	// retention then we will need to check the ack floor of all consumers.
+	if ocfg.Retention != cfg.Retention && cfg.Retention == InterestPolicy {
+		toUpdate := make([]*consumer, 0, len(mset.consumers))
+		for _, c := range mset.consumers {
+			if c.retention != cfg.Retention {
+				toUpdate = append(toUpdate, c)
+			}
+		}
+		mset.mu.Unlock()
+		for _, c := range toUpdate {
+			c.mu.Lock()
+			c.retention = cfg.Retention
+			c.mu.Unlock()
+			if c.retention == InterestPolicy {
+				c.checkStateForInterestStream()
```

Comment:
Nice! I saw this and was confused for a second then realized what you did here..